src/protocol/attribute.rs

```diff
@@ -0,0 +1,224 @@
+    ctx: &'a ProtocolContext<'a, G, AttributionModel>,
+    step: AttributionModel,
+}
+
+impl<'a, G> Attribute<'a, G> {
+    const _IS_TRIGGER_BIT: usize = 2;
+    const _HELPER_BIT: usize = 3;
+    const _VALUE: usize = 5;
+
+    #[allow(dead_code)]
+    pub fn new(ctx: &'a ProtocolContext<'a, G, AttributionModel>, step: AttributionModel) -> Self {
+        Self { ctx, step }
+    }
+
+    /// Executes the protocol and produces an attribution result for the model specified by `step` in `Attribute::new`.
+    /// Attribution result will be used to calculate the aggregation in the next step.
+    #[allow(dead_code)]
+    pub async fn execute<M, F>(
+        &self,
+        input: &'a [Vec<Replicated<F>>],
+    ) -> Result<Vec<Replicated<F>>, BoxError>
+    where
+        M: Mesh,
+        G: Gateway<M, AttributionModel>,
+        F: Field,
+    {
+        match self.step {
+            AttributionModel::LastTouch => self.last_touch_attribution(input).await,
+        }
+    }
+
+    async fn last_touch_attribution<M, F>(
+        &self,
+        input: &'a [Vec<Replicated<F>>],
+    ) -> Result<Vec<Replicated<F>>, BoxError>
+    where
+        M: Mesh,
+        G: Gateway<M, AttributionModel>,
+        F: Field,
+    {
+        // TOOD: input length range check. Can we assume `assert!(num_rows < u32::MAX)`?
+        let num_rows = input[0].len();
+
+        // 1. Create Credit and Stop bit vectors
+        let mut credits: Vec<Replicated<F>> = Vec::with_capacity(num_rows);
+        input[Self::_VALUE].iter().for_each(|x| credits.push(*x));
+
+        let one = self.ctx.gateway.get_channel(self.step).share_of_one::<F>();
+        let mut stop_bits: Vec<Replicated<F>> = vec![one; num_rows];
+
+        // 2. Accumulate
+        //
+        // Last touch accumulation algorithm:
+        //     b = stop_bit * successor.helper_bit * successor.trigger_bit;
+        //     credit = credit + b * successor.credit;
+        //     stop_bit = b * successor.stop_bit;
+        let mut step_size = 1;
+        while step_size < num_rows {
+            let end = num_rows - step_size;
+
+            // first, calculate [successor.helper_bit * successor.trigger_bit]
+            let mut b = self
+                .multiply_all(
+                    &input[Self::_HELPER_BIT][step_size..],
+                    &input[Self::_IS_TRIGGER_BIT][step_size..],
+                )
+                .await?;
+
+            // since `stop_bits` is initialized with `1`s, we only multiply [stop_bit] on the second and later iterations
+            if step_size > 1 {
+                b = self.multiply_all(&stop_bits[..end], &b).await?;
+            }
+
+            // [b * successor.credit]
+            let b_cred = self.multiply_all(&b[..end], &credits[step_size..]).await?;
+
+            // [credit] = [credit + b * successor.credit]
+            b_cred
+                .into_iter()
+                .enumerate()
+                .for_each(|(i, x)| credits[i] += x);
```

Comment:
OK, I see what you're doing here. It's very subtle. The index "i" refers to the correct successor here because the b_cred vector is shifted compared to the original input.

That is very clever, but I'm worried its a bit too subtle. It's harder to understand and I worry will lead to bugs.

Honestly, I would rather see something like:

b_cred = foo[i] + bar[i + step_size];

I feel like that makes it easier to see what's going on.